ext/random/engine_mt19937.c

```diff
@@ -0,0 +1,339 @@
+   |          Sascha Schumann <sascha@schumann.cx>                        |
+   |          Pedro Melo <melo@ip.pt>                                     |
+   |          Sterling Hughes <sterling@php.net>                          |
+   |          Sammy Kaye Powers <me@sammyk.me>                            |
+   |          Go Kudo <g-kudo@colopl.co.jp>                               |
+   |                                                                      |
+   | Based on code from: Richard J. Wagner <rjwagner@writeme.com>         |
+   |                     Makoto Matsumoto <matumoto@math.keio.ac.jp>      |
+   |                     Takuji Nishimura                                 |
+   |                     Shawn Cokus <Cokus@math.washington.edu>          |
+   |                     David Blackman                                   |
+   |                     Sebastiano Vigna <vigna@acm.org>                 |
+   |                     Melissa O'Neill <oneill@pcg-random.org>          |
+   +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "php.h"
+#include "php_random.h"
+
+#include "ext/spl/spl_exceptions.h"
+#include "Zend/zend_exceptions.h"
+
+#define N             MT_N                 /* length of state vector */
+#define M             (397)                /* a period parameter */
+#define hiBit(u)      ((u) & 0x80000000U)  /* mask all but highest   bit of u */
+#define loBit(u)      ((u) & 0x00000001U)  /* mask all but lowest    bit of u */
+#define loBits(u)     ((u) & 0x7FFFFFFFU)  /* mask     the highest   bit of u */
+#define mixBits(u, v) (hiBit(u)|loBits(v)) /* move hi bit of u to hi bit of v */
+
+#define twist(m,u,v)  (m ^ (mixBits(u,v)>>1) ^ ((uint32_t)(-(int32_t)(loBit(v))) & 0x9908b0dfU))
+#define twist_php(m,u,v)  (m ^ (mixBits(u,v)>>1) ^ ((uint32_t)(-(int32_t)(loBit(u))) & 0x9908b0dfU))
+
+static inline void mt19937_reload(php_random_status_state_mt19937 *state) {
+	uint32_t *p = state->state;
+	int i;
+
+	if (state->mode == MT_RAND_MT19937) {
+		for (i = N - M; i--; ++p) {
+			*p = twist(p[M], p[0], p[1]);
+		}
+		for (i = M; --i; ++p) {
+			*p = twist(p[M-N], p[0], p[1]);
+		}
+		*p = twist(p[M-N], p[0], state->state[0]);
+	} else {
+		for (i = N - M; i--; ++p) {
+			*p = twist_php(p[M], p[0], p[1]);
+		}
+		for (i = M; --i; ++p) {
+			*p = twist_php(p[M-N], p[0], p[1]);
+		}
+		*p = twist_php(p[M-N], p[0], state->state[0]);
+	}
+
+	state->count = 0;
+}
+
+static inline void mt19937_seed_state(php_random_status_state_mt19937 *state, uint64_t seed)
+{
+	/* Initialize generator state with seed
+	   See Knuth TAOCP Vol 2, 3rd Ed, p.106 for multiplier.
+	   In previous versions, most significant bits (MSBs) of the seed affect
+	   only MSBs of the state array.  Modified 9 Jan 2002 by Makoto Matsumoto. */
+	state->state[0] = seed & 0xffffffffU;
+	for (state->count = 1; state->count < MT_N; state->count++) {
```

Comment:
why don't you use local variable here and update the count after the loop? Dereferencing in the loop is not the most effective...

Reply:
fixed: 0708741909c531fc3ac5681c7aba70ec7f1ae048